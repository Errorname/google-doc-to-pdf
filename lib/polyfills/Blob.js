"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_fetch_1 = require("node-fetch");
const realBlob = typeof window !== 'undefined' ? window.Blob : node_fetch_1.Blob;
exports.default = realBlob;
