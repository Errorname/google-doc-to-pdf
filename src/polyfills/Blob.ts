import { Blob as BlobNode } from 'node-fetch'

const realBlob = typeof window !== 'undefined' ? window.Blob : BlobNode

export default realBlob
