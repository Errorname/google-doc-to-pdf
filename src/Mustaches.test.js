import Mustaches from './Mustaches'
import Blob from './polyfills/Blob'
import crossFetch from 'cross-fetch'

jest.mock('cross-fetch', () => jest.fn())
jest.mock('./polyfills/Blob')
Blob.mockImplementation(args => ({ args: args.join('') }))

const buildFetchImpl = crossFetch => {
  const json = val => Promise.resolve({ status: 200, json: () => val })

  crossFetch.mockImplementation(url => {
    // drive.get
    if (url.includes('?fields=parents')) {
      return json({ parents: ['parent-id-123'] })
    }
    // drive.copy
    else if (url.endsWith('/copy')) {
      return json({ id: 'new-file-id-123' })
    }
    // drive.export
    else if (url.includes('export?mimeType=')) {
      return Promise.resolve({ status: 200, blob: () => 'BLOB123' })
    }
    // drive.create
    else if (url.includes('?uploadType=')) {
      return json({ id: 'upload-id-123' })
    }
    // docs.update
    else if (url.includes(':batchUpdate')) {
      return json({ status: 'OK' })
    }
    // docs.get
    else if (url.includes('v1/documents')) {
      return json({
        body: {
          content: [
            {
              paragraph: {
                elements: [
                  {
                    textRun: {
                      content:
                        'Hello {{ name | smurf }}! Do you like {{ movies[0].title | lowercase }}?'
                    },
                    startIndex: 0,
                    endIndex: 71
                  }
                ]
              }
            }
          ]
        }
      })
    }
  })
}

describe('mustaches', () => {
  const token = () => 'my-token'

  let mustaches

  beforeEach(() => {
    crossFetch.mockReset()
    buildFetchImpl(crossFetch)

    mustaches = new Mustaches({ token })
  })

  test('nominal use case', async () => {
    await mustaches.interpolate({
      source: 'source-id-123',
      destination: 'destination-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('no destination given', async () => {
    await mustaches.interpolate({
      source: 'source-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('with name', async () => {
    await mustaches.interpolate({
      source: 'source-id-123',
      destination: 'destination-id-123',
      name: 'New interpolated file',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('with custom formatters ', async () => {
    await mustaches.interpolate({
      source: 'source-id-123',
      destination: 'destination-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      },
      formatters: {
        smurf: () => 'Smurf'
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('export', async () => {
    await mustaches.interpolate({
      source: 'source-id-123',
      destination: 'destination-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      },
      export: 'application/pdf'
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('discovery', async () => {
    const placeholders = await mustaches.discovery({
      source: 'source-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
    expect(placeholders).toMatchSnapshot()
  })

  test('discovery with no data', async () => {
    const placeholders = await mustaches.discovery({
      source: 'source-id-123'
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
    expect(placeholders).toMatchSnapshot()
  })

  test('discovery with formatters', async () => {
    const placeholders = await mustaches.discovery({
      source: 'source-id-123',
      data: {
        name: 'Thibaud',
        movies: [{ title: 'Lost in Translation' }]
      },
      formatters: {
        smurf: () => 'Smurf'
      }
    })

    expect(crossFetch.mock.calls).toMatchSnapshot()
    expect(placeholders).toMatchSnapshot()
  })
})
