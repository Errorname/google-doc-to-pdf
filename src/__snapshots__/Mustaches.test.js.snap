// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mustaches discovery 1`] = `
Array [
  Array [
    "https://docs.googleapis.com/v1/documents/source-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`mustaches discovery 2`] = `
Array [
  Object {
    "input": Object {
      "raw": "name",
      "value": "Thibaud",
    },
    "output": "Thibaud",
    "pipes": Array [
      Object {
        "error": [Error: smurf is not defined.],
        "output": "Thibaud",
        "raw": "smurf",
      },
    ],
    "position": Object {
      "end": 24,
      "start": 6,
    },
    "raw": "{{ name | smurf }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movies[0].title",
      "value": "Lost in Translation",
    },
    "output": "lost in translation",
    "pipes": Array [
      Object {
        "output": "lost in translation",
        "raw": "lowercase",
      },
    ],
    "position": Object {
      "end": 71,
      "start": 38,
    },
    "raw": "{{ movies[0].title | lowercase }}",
    "type": "content",
  },
]
`;

exports[`mustaches discovery with formatters 1`] = `
Array [
  Array [
    "https://docs.googleapis.com/v1/documents/source-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`mustaches discovery with formatters 2`] = `
Array [
  Object {
    "input": Object {
      "raw": "name",
      "value": "Thibaud",
    },
    "output": "Smurf",
    "pipes": Array [
      Object {
        "output": "Smurf",
        "raw": "smurf",
      },
    ],
    "position": Object {
      "end": 24,
      "start": 6,
    },
    "raw": "{{ name | smurf }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movies[0].title",
      "value": "Lost in Translation",
    },
    "output": "lost in translation",
    "pipes": Array [
      Object {
        "output": "lost in translation",
        "raw": "lowercase",
      },
    ],
    "position": Object {
      "end": 71,
      "start": 38,
    },
    "raw": "{{ movies[0].title | lowercase }}",
    "type": "content",
  },
]
`;

exports[`mustaches discovery with no data 1`] = `
Array [
  Array [
    "https://docs.googleapis.com/v1/documents/source-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`mustaches discovery with no data 2`] = `
Array [
  Object {
    "input": Object {
      "error": [Error: name is undefined.],
      "raw": "name",
      "value": "",
    },
    "output": "",
    "pipes": Array [
      Object {
        "error": [Error: smurf is not defined.],
        "output": "",
        "raw": "smurf",
      },
    ],
    "position": Object {
      "end": 24,
      "start": 6,
    },
    "raw": "{{ name | smurf }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "error": [Error: movies is undefined.],
      "raw": "movies[0].title",
      "value": "",
    },
    "output": "",
    "pipes": Array [
      Object {
        "output": "",
        "raw": "lowercase",
      },
    ],
    "position": Object {
      "end": 71,
      "start": 38,
    },
    "raw": "{{ movies[0].title | lowercase }}",
    "type": "content",
  },
]
`;

exports[`mustaches export 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123?fields=parents",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/export?mimeType=application/pdf",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&alt=json",
    Object {
      "body": "{\\"args\\":\\"----BOUNDARY\\\\nContent-Type: application/json; charset=UTF-8\\\\n\\\\n{\\\\\\"name\\\\\\":\\\\\\"Export\\\\\\",\\\\\\"parents\\\\\\":[\\\\\\"parent-id-123\\\\\\"]}\\\\n\\\\n----BOUNDARY\\\\nContent-Type: application/pdf\\\\nContent-Encoding: base64\\\\n\\\\nBLOB123\\\\n\\\\n----BOUNDARY--\\"}",
      "headers": Object {
        "Accept": "application/json",
        "Content-Type": "multipart/related; boundary=--BOUNDARY",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches export with destination 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/export?mimeType=application/pdf",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&alt=json",
    Object {
      "body": "{\\"args\\":\\"----BOUNDARY\\\\nContent-Type: application/json; charset=UTF-8\\\\n\\\\n{\\\\\\"name\\\\\\":\\\\\\"Export\\\\\\",\\\\\\"parents\\\\\\":[\\\\\\"destination-id-123\\\\\\"]}\\\\n\\\\n----BOUNDARY\\\\nContent-Type: application/pdf\\\\nContent-Encoding: base64\\\\n\\\\nBLOB123\\\\n\\\\n----BOUNDARY--\\"}",
      "headers": Object {
        "Accept": "application/json",
        "Content-Type": "multipart/related; boundary=--BOUNDARY",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches export with name 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123?fields=parents",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/export?mimeType=application/pdf",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&alt=json",
    Object {
      "body": "{\\"args\\":\\"----BOUNDARY\\\\nContent-Type: application/json; charset=UTF-8\\\\n\\\\n{\\\\\\"name\\\\\\":\\\\\\"My great pdf\\\\\\",\\\\\\"parents\\\\\\":[\\\\\\"parent-id-123\\\\\\"]}\\\\n\\\\n----BOUNDARY\\\\nContent-Type: application/pdf\\\\nContent-Encoding: base64\\\\n\\\\nBLOB123\\\\n\\\\n----BOUNDARY--\\"}",
      "headers": Object {
        "Accept": "application/json",
        "Content-Type": "multipart/related; boundary=--BOUNDARY",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches interpolate 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/copy",
    Object {
      "body": "{\\"parents\\":[\\"destination-id-123\\"]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123:batchUpdate",
    Object {
      "body": "{\\"documentId\\":\\"new-file-id-123\\",\\"requests\\":[{\\"replaceAllText\\":{\\"replaceText\\":\\"Thibaud\\",\\"containsText\\":{\\"text\\":\\"{{ name | smurf }}\\",\\"matchCase\\":true}}},{\\"replaceAllText\\":{\\"replaceText\\":\\"lost in translation\\",\\"containsText\\":{\\"text\\":\\"{{ movies[0].title | lowercase }}\\",\\"matchCase\\":true}}}]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches interpolate with custom formatters  1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/copy",
    Object {
      "body": "{\\"parents\\":[\\"destination-id-123\\"]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123:batchUpdate",
    Object {
      "body": "{\\"documentId\\":\\"new-file-id-123\\",\\"requests\\":[{\\"replaceAllText\\":{\\"replaceText\\":\\"Smurf\\",\\"containsText\\":{\\"text\\":\\"{{ name | smurf }}\\",\\"matchCase\\":true}}},{\\"replaceAllText\\":{\\"replaceText\\":\\"lost in translation\\",\\"containsText\\":{\\"text\\":\\"{{ movies[0].title | lowercase }}\\",\\"matchCase\\":true}}}]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches interpolate with name 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/copy",
    Object {
      "body": "{\\"parents\\":[\\"destination-id-123\\"],\\"name\\":\\"New interpolated file\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123:batchUpdate",
    Object {
      "body": "{\\"documentId\\":\\"new-file-id-123\\",\\"requests\\":[{\\"replaceAllText\\":{\\"replaceText\\":\\"Thibaud\\",\\"containsText\\":{\\"text\\":\\"{{ name | smurf }}\\",\\"matchCase\\":true}}},{\\"replaceAllText\\":{\\"replaceText\\":\\"lost in translation\\",\\"containsText\\":{\\"text\\":\\"{{ movies[0].title | lowercase }}\\",\\"matchCase\\":true}}}]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`mustaches interpolate with no destination given 1`] = `
Array [
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123?fields=parents",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://www.googleapis.com/drive/v3/files/source-id-123/copy",
    Object {
      "body": "{\\"parents\\":[\\"parent-id-123\\"]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
  Array [
    "https://docs.googleapis.com/v1/documents/new-file-id-123:batchUpdate",
    Object {
      "body": "{\\"documentId\\":\\"new-file-id-123\\",\\"requests\\":[{\\"replaceAllText\\":{\\"replaceText\\":\\"Thibaud\\",\\"containsText\\":{\\"text\\":\\"{{ name | smurf }}\\",\\"matchCase\\":true}}},{\\"replaceAllText\\":{\\"replaceText\\":\\"lost in translation\\",\\"containsText\\":{\\"text\\":\\"{{ movies[0].title | lowercase }}\\",\\"matchCase\\":true}}}]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;
